### PR TITLE
[OpInfo] Add expected_failure kwarg to SkipInfo

### LINKED ISSUE #62354: Some way to specify expected failures for OpInfo-based tests
## 🚀 Feature

A way to specify expected failure for OpInfo-based tests that works with unittest's expected failure reporting

## Motivation

For functorch we have automated coverage tests for OpInfos. E.g. there is a test that for each OpInfo, checks vmap support for the OpInfo. Tests that are currently failing are put into a denylist and skipped instead of running. It would be nice if we could mark them as expected failure instead.

## Alternatives

We can emulate expected failures by doing a try-except and checking for the presence of an exception. The main problem I have with this approach is the UX: the unittest runner shows an "expected failure" as a passed test, when it would be nice to see that it is an expected failure in the runner.

cc @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62725
* __->__ #62963
* #62713

Differential Revision: [D30327199](https://our.internmc.facebook.com/intern/diff/D30327199)

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -74,11 +74,22 @@ class SkipInfo(DecorateInfo):
...
-    def __init__(self, cls_name=None, test_name=None, *,
-                 device_type=None, dtypes=None, active_if=True):
-        super().__init__(decorators=skipIf(True, "Skipped!"), cls_name=cls_name,
-                         test_name=test_name, device_type=device_type, dtypes=dtypes,
-                         active_if=active_if)
+    def __init__(
+            self, cls_name=None, test_name=None, *, device_type=None, dtypes=None, active_if=True,
+            expected_failure=False):
+        """
+        Args:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:81]: Nice docs update

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -74,11 +74,22 @@ class SkipInfo(DecorateInfo):
...
+            cls_name: the name of the test class to skip
+            test_name: the name of the test within the test class to skip
+            device_type: the devices for which to skip the tests
+            dtypes: the dtypes for which to skip the tests
+            active_if: whether tests matching the above arguments should be skipped
+            expected_failure: whether to assert that skipped tests fail
+        """
+        decorator = expectedFailure(device_type) if expected_failure else skipIf(True, "Skipped!")
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:89]: It's a little weird that the expectedFailure decorator constructor requires a device_type although it allows None. But I guess that's an issue for another time

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30327199).

**Pull request closed without merging**